# Telecom Churn Prediction

In [1]:
# import the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Reading and Understanding

In [2]:
# load the dataset
churn_df = pd.read_csv('telecom_churn_data.csv')
churn_df.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6  ...  sachet_3g_9  fb_user_6  fb_user_7  \
0            9/30/2014  197.385  ...            0        1.0        1.0   
1            9/30/2014   34.047  ...            0        NaN        1.0   
2            9/30/2014  167.690  ...            0        NaN        NaN   
3            9/30/2014  221.338  ...            0        NaN        NaN   
4            9/30/2014  261.636  ...            0        0.0        NaN   

   fb_user_8  fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  
0        1.0        NaN   968        30.4         0.0      101.20        3.58  
1        1.0        NaN  1006         0.0         0.0        0.00        0.00  
2        NaN        1.0  1103         0.0         0.0        4.17        0.00  
3        NaN        NaN  2491         0.0         0.0        0.00        0.00  
4        NaN        NaN  1526         0.0         0.0        0.00        0.00  

[5 rows x 226 columns]

In [3]:
churn_df.shape

(99999, 226)

In [4]:
# see basic imformation about the dataset
churn_df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 226 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    mobile_number             int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    last_date_of_month_9      object 
 9    arpu_6                    float64
 10   arpu_7                    float64
 11   arpu_8                    float64
 12   arpu_9                    float64
 13   onnet_mou_6               float64
 14   onnet_mou_7               float64
 15   onnet_mou_8               float64
 16   onnet_mou_9               float64
 17   offnet_mou_6              float64
 18   offnet_mou_7              float64
 19   offnet_mou_8              float64
 20   offn

In [5]:
# function for printing columns with their percentage missing values
def print_missing_val_percent_in_cols(df):
    for col in df.columns:
        missing_percent = df[col].isnull().sum() / df.shape[0] * 100
        print('{}: {}'.format(col, round(missing_percent, 2)))
        
# function to return the columns with missing values
def find_missing_val_cols(df):
    missing_val_cols = []
    for col in df.columns:
        if df[col].isnull().sum() > 0:
            missing_val_cols.append(col)
    return missing_val_cols

In [6]:
# checking the percentage of missing values in the data set
print_missing_val_percent_in_cols(churn_df)

mobile_number: 0.0
circle_id: 0.0
loc_og_t2o_mou: 1.02
std_og_t2o_mou: 1.02
loc_ic_t2o_mou: 1.02
last_date_of_month_6: 0.0
last_date_of_month_7: 0.6
last_date_of_month_8: 1.1
last_date_of_month_9: 1.66
arpu_6: 0.0
arpu_7: 0.0
arpu_8: 0.0
arpu_9: 0.0
onnet_mou_6: 3.94
onnet_mou_7: 3.86
onnet_mou_8: 5.38
onnet_mou_9: 7.75
offnet_mou_6: 3.94
offnet_mou_7: 3.86
offnet_mou_8: 5.38
offnet_mou_9: 7.75
roam_ic_mou_6: 3.94
roam_ic_mou_7: 3.86
roam_ic_mou_8: 5.38
roam_ic_mou_9: 7.75
roam_og_mou_6: 3.94
roam_og_mou_7: 3.86
roam_og_mou_8: 5.38
roam_og_mou_9: 7.75
loc_og_t2t_mou_6: 3.94
loc_og_t2t_mou_7: 3.86
loc_og_t2t_mou_8: 5.38
loc_og_t2t_mou_9: 7.75
loc_og_t2m_mou_6: 3.94
loc_og_t2m_mou_7: 3.86
loc_og_t2m_mou_8: 5.38
loc_og_t2m_mou_9: 7.75
loc_og_t2f_mou_6: 3.94
loc_og_t2f_mou_7: 3.86
loc_og_t2f_mou_8: 5.38
loc_og_t2f_mou_9: 7.75
loc_og_t2c_mou_6: 3.94
loc_og_t2c_mou_7: 3.86
loc_og_t2c_mou_8: 5.38
loc_og_t2c_mou_9: 7.75
loc_og_mou_6: 3.94
loc_og_mou_7: 3.86
loc_og_mou_8: 5.38
loc_og_mou_9: 7.7

In [7]:
missing_val_cols = find_missing_val_cols(churn_df)
print(missing_val_cols)

['loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou', 'last_date_of_month_7', 'last_date_of_month_8', 'last_date_of_month_9', 'onnet_mou_6', 'onnet_mou_7', 'onnet_mou_8', 'onnet_mou_9', 'offnet_mou_6', 'offnet_mou_7', 'offnet_mou_8', 'offnet_mou_9', 'roam_ic_mou_6', 'roam_ic_mou_7', 'roam_ic_mou_8', 'roam_ic_mou_9', 'roam_og_mou_6', 'roam_og_mou_7', 'roam_og_mou_8', 'roam_og_mou_9', 'loc_og_t2t_mou_6', 'loc_og_t2t_mou_7', 'loc_og_t2t_mou_8', 'loc_og_t2t_mou_9', 'loc_og_t2m_mou_6', 'loc_og_t2m_mou_7', 'loc_og_t2m_mou_8', 'loc_og_t2m_mou_9', 'loc_og_t2f_mou_6', 'loc_og_t2f_mou_7', 'loc_og_t2f_mou_8', 'loc_og_t2f_mou_9', 'loc_og_t2c_mou_6', 'loc_og_t2c_mou_7', 'loc_og_t2c_mou_8', 'loc_og_t2c_mou_9', 'loc_og_mou_6', 'loc_og_mou_7', 'loc_og_mou_8', 'loc_og_mou_9', 'std_og_t2t_mou_6', 'std_og_t2t_mou_7', 'std_og_t2t_mou_8', 'std_og_t2t_mou_9', 'std_og_t2m_mou_6', 'std_og_t2m_mou_7', 'std_og_t2m_mou_8', 'std_og_t2m_mou_9', 'std_og_t2f_mou_6', 'std_og_t2f_mou_7', 'std_og_t2f_mou_8', 'std_og_t

#### **Business Decision: The missing values for data recharge amount, call recharge amount and avergae recharge amount can be imputed with 0**

In [15]:
churn_df[['total_rech_data_6', 'total_rech_data_7', 
          'total_rech_data_8', 'total_rech_data_9',
          'av_rech_amt_data_6', 'av_rech_amt_data_7',
          'av_rech_amt_data_8', 'av_rech_amt_data_9',
          'total_rech_amt_6', 'total_rech_amt_7',
          'total_rech_amt_8', 'total_rech_amt_9']] = churn_df[['total_rech_data_6', 'total_rech_data_7',
                                                               'total_rech_data_8', 'total_rech_data_9',
                                                               'av_rech_amt_data_6', 'av_rech_amt_data_7',
                                                               'av_rech_amt_data_8', 'av_rech_amt_data_9',
                                                               'total_rech_amt_6', 'total_rech_amt_7',
                                                               'total_rech_amt_8', 'total_rech_amt_9']].fillna(0)

#### Deriving the column that has total data recharge amount for every month

In [25]:
# creating total amount spend on data for each month
churn_df['total_rech_data_amt_6'] = churn_df['total_rech_data_6'] * churn_df['av_rech_amt_data_6']
churn_df['total_rech_data_amt_7'] = churn_df['total_rech_data_7'] * churn_df['av_rech_amt_data_7']
churn_df['total_rech_data_amt_8'] = churn_df['total_rech_data_8'] * churn_df['av_rech_amt_data_8']
churn_df['total_rech_data_amt_9'] = churn_df['total_rech_data_9'] * churn_df['av_rech_amt_data_9']

In [26]:
churn_df.shape

(99999, 230)

#### Deriving the columns that has the total recharge amount for calling plus data included for every month

In [27]:
# creating total amound spend every month including calls and data
churn_df['total_rech_call_plus_data_6'] = churn_df['total_rech_amt_6'] + churn_df['total_rech_data_amt_6']
churn_df['total_rech_call_plus_data_7'] = churn_df['total_rech_amt_7'] + churn_df['total_rech_data_amt_7']
churn_df['total_rech_call_plus_data_8'] = churn_df['total_rech_amt_8'] + churn_df['total_rech_data_amt_8']
churn_df['total_rech_call_plus_data_9'] = churn_df['total_rech_amt_9'] + churn_df['total_rech_data_amt_9']

#### **Business Logic: Finding the average amount of total recharge of call plus data for the good phase i.e. the months 6 and 7**

In [30]:
# creating average amount spend on call plus data in good phase
churn_df['total_rech_call_plus_data_good_phase'] = (churn_df['total_rech_call_plus_data_6'] + churn_df['total_rech_call_plus_data_7']) / 2

#### **Business Decision: Only the customers that have average recharge amount more than 70 percentile during the good phase (i.e months 6 and 7) are considered for the churn prediction as according to the businees logic these are the customers that account for 80% of the total revenue**

In [33]:
# computing the 70% quantile recharge amount
high_value_cust_rech_amt = churn_df['total_rech_call_plus_data_good_phase'].quantile(0.7)
print(high_value_cust_rech_amt)

478.0


#### Observation: The threshold recharge amount for deciding a customer as high value comes out to be Rs 478 per month during the good phase

In [36]:
# finding the high value customers
churn_df = churn_df[churn_df['total_rech_call_plus_data_good_phase'] > high_value_cust_rech_amt]
churn_df.shape

(29953, 235)

#### NOTE: The data set that contains only the high value customers has roughly 29.9k records

### Calculating the Churn label for the month of September

In [41]:
# finding the usage-based churn for month 9
#finding the total call and data usage for month 9
churn_df['total_call_data_usage_9'] = churn_df['total_ic_mou_9'] + churn_df['total_og_mou_9'] + churn_df['vol_2g_mb_9'] + churn_df['vol_3g_mb_9']

# finding the Churn label on usage
churn_df['Churn'] = (churn_df['total_call_data_usage_9'] > 0).map({True: 0, False: 1})

/tmp/ipykernel_9259/3092612585.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  churn_df['total_call_data_usage_9'] = churn_df['total_ic_mou_9'] + churn_df['total_og_mou_9'] + churn_df['vol_2g_mb_9'] + churn_df['vol_3g_mb_9']
/tmp/ipykernel_9259/3092612585.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  churn_df['Churn'] = (churn_df['total_call_data_usage_9'] > 0).map({True: 0, False: 1})


In [42]:
churn_df.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0      7000842753        109             0.0             0.0             0.0   
7      7000701601        109             0.0             0.0             0.0   
8      7001524846        109             0.0             0.0             0.0   
21     7002124215        109             0.0             0.0             0.0   
23     7000887461        109             0.0             0.0             0.0   

   last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0             6/30/2014            7/31/2014            8/31/2014   
7             6/30/2014            7/31/2014            8/31/2014   
8             6/30/2014            7/31/2014            8/31/2014   
21            6/30/2014            7/31/2014            8/31/2014   
23            6/30/2014            7/31/2014            8/31/2014   

   last_date_of_month_9    arpu_6  ...  total_rech_data_amt_7  \
0             9/30/2014   197.385  ...                  252.0   
7             9/30/2014  1069.180  ...                    0.0   
8             9/30/2014   378.721  ...                  354.0   
21            9/30/2014   514.453  ...                    0.0   
23            9/30/2014    74.350  ...                  712.0   

    total_rech_data_amt_8  total_rech_data_amt_9  total_rech_call_plus_data_6  \
0                   252.0                    0.0                        614.0   
7                     0.0                    0.0                       1580.0   
8                   207.0                    0.0                        437.0   
21                    0.0                    0.0                        600.0   
23                  540.0                  252.0                          0.0   

    total_rech_call_plus_data_7  total_rech_call_plus_data_8  \
0                         504.0                        504.0   
7                         790.0                       3638.0   
8                         955.0                        327.0   
21                        680.0                        718.0   
23                       1166.0                        979.0   

    total_rech_call_plus_data_9  total_rech_call_plus_data_good_phase  \
0                           0.0                                 559.0   
7                           0.0                                1185.0   
8                         186.0                                 696.0   
21                        680.0                                 640.0   
23                        504.0                                 583.0   

    total_call_data_usage_9  Churn  
0                      0.00      1  
7                      0.00      1  
8                    177.09      0  
21                  1284.14      0  
23                   583.78      0  

[5 rows x 237 columns]

#### Dropping the columns with month 9 data as it won't  be available at the time of prediction

In [44]:
# finding month 9 columns 
month_9_cols = []
for col in churn_df.columns:
    if '_9' in col:
        month_9_cols.append(col)

# dropping month 9 columns
churn_df.drop(month_9_cols, axis=1, inplace=True)

/home/sameersoin/anaconda3/envs/data-science/lib/python3.9/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [45]:
churn_df.shape

(29953, 180)

### Exploring the object columns

In [49]:
object_data_type_cols = churn_df.select_dtypes('object')
object_data_type_cols

last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0                6/30/2014            7/31/2014            8/31/2014   
7                6/30/2014            7/31/2014            8/31/2014   
8                6/30/2014            7/31/2014            8/31/2014   
21               6/30/2014            7/31/2014            8/31/2014   
23               6/30/2014            7/31/2014            8/31/2014   
...                    ...                  ...                  ...   
99981            6/30/2014            7/31/2014            8/31/2014   
99984            6/30/2014            7/31/2014            8/31/2014   
99986            6/30/2014            7/31/2014            8/31/2014   
99988            6/30/2014            7/31/2014            8/31/2014   
99997            6/30/2014            7/31/2014            8/31/2014   

      date_of_last_rech_6 date_of_last_rech_7 date_of_last_rech_8  \
0               6/21/2014           7/16/2014            8/8/2014   
7               6/27/2014           7/25/2014           8/26/2014   
8               6/25/2014           7/31/2014           8/30/2014   
21              6/30/2014           7/31/2014           8/31/2014   
23              6/18/2014            7/7/2014           8/24/2014   
...                   ...                 ...                 ...   
99981           6/17/2014           7/25/2014           8/27/2014   
99984           6/21/2014           7/31/2014           8/27/2014   
99986           6/20/2014           7/28/2014           8/18/2014   
99988           6/30/2014           7/28/2014           8/29/2014   
99997           6/17/2014           7/19/2014           8/20/2014   

      date_of_last_rech_data_6 date_of_last_rech_data_7  \
0                    6/21/2014                7/16/2014   
7                          NaN                      NaN   
8                          NaN                7/31/2014   
21                         NaN                      NaN   
23                         NaN                 7/7/2014   
...                        ...                      ...   
99981                6/17/2014                7/14/2014   
99984                6/15/2014                7/31/2014   
99986                6/17/2014                7/18/2014   
99988                      NaN                7/26/2014   
99997                6/17/2014                7/19/2014   

      date_of_last_rech_data_8  
0                     8/8/2014  
7                          NaN  
8                    8/23/2014  
21                         NaN  
23                    8/6/2014  
...                        ...  
99981                8/23/2014  
99984                8/27/2014  
99986                8/14/2014  
99988                8/23/2014  
99997                8/20/2014  

[29953 rows x 9 columns]

**Observation: All the object data type columns are dates**

In [54]:
# converting the object type columns to datetime
for col in object_data_type_cols:
    churn_df[col] = churn_df[col].apply(pd.to_datetime)

In [56]:
churn_df

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
0         7000842753        109             0.0             0.0   
7         7000701601        109             0.0             0.0   
8         7001524846        109             0.0             0.0   
21        7002124215        109             0.0             0.0   
23        7000887461        109             0.0             0.0   
...              ...        ...             ...             ...   
99981     7000630859        109             0.0             0.0   
99984     7000661676        109             0.0             0.0   
99986     7001729035        109             0.0             0.0   
99988     7002111859        109             0.0             0.0   
99997     7000498689        109             0.0             0.0   

       loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
0                 0.0           2014-06-30           2014-07-31   
7                 0.0           2014-06-30           2014-07-31   
8                 0.0           2014-06-30           2014-07-31   
21                0.0           2014-06-30           2014-07-31   
23                0.0           2014-06-30           2014-07-31   
...               ...                  ...                  ...   
99981             0.0           2014-06-30           2014-07-31   
99984             0.0           2014-06-30           2014-07-31   
99986             0.0           2014-06-30           2014-07-31   
99988             0.0           2014-06-30           2014-07-31   
99997             0.0           2014-06-30           2014-07-31   

      last_date_of_month_8    arpu_6    arpu_7  ...  jun_vbc_3g  sep_vbc_3g  \
0               2014-08-31   197.385   214.816  ...      101.20        3.58   
7               2014-08-31  1069.180  1349.850  ...       18.74        0.00   
8               2014-08-31   378.721   492.223  ...      122.16        0.00   
21              2014-08-31   514.453   597.753  ...        0.00        0.00   
23              2014-08-31    74.350   193.897  ...        0.00        0.00   
...                    ...       ...       ...  ...         ...         ...   
99981           2014-08-31   384.316   255.405  ...      668.05        0.33   
99984           2014-08-31   328.594   202.966  ...        0.00        0.00   
99986           2014-08-31   644.973   455.228  ...      604.08        0.00   
99988           2014-08-31   312.558   512.932  ...        0.00        0.00   
99997           2014-08-31   322.991   303.386  ...     1173.18        0.00   

       total_rech_data_amt_6  total_rech_data_amt_7  total_rech_data_amt_8  \
0                      252.0                  252.0                  252.0   
7                        0.0                    0.0                    0.0   
8                        0.0                  354.0                  207.0   
21                       0.0                    0.0                    0.0   
23                       0.0                  712.0                  540.0   
...                      ...                    ...                    ...   
99981                  252.0                  252.0                  252.0   
99984                  154.0                  484.0                  585.0   
99986                   17.0                   14.0                   25.0   
99988                    0.0                  154.0                  154.0   
99997                 1749.0                  716.0                 2864.0   

       total_rech_call_plus_data_6  total_rech_call_plus_data_7  \
0                            614.0                        504.0   
7                           1580.0                        790.0   
8                            437.0                        955.0   
21                           600.0                        680.0   
23                             0.0                       1166.0   
...                            ...                          ...   
99981                        504.0        

**Object data type columns are converted to datetime type**

#### Deriving the number of days since last recharge column

In [58]:
# finding the number of days since last call recharge for every month
churn_df['num_of_days_last_rech_6'] = (churn_df['last_date_of_month_6'] - churn_df['date_of_last_rech_6']).dt.days
churn_df['num_of_days_last_rech_7'] = (churn_df['last_date_of_month_7'] - churn_df['date_of_last_rech_7']).dt.days
churn_df['num_of_days_last_rech_8'] = (churn_df['last_date_of_month_8'] - churn_df['date_of_last_rech_8']).dt.days

# finding the number of days since last data recharge for every month
churn_df['num_of_days_last_rech_data_6'] = (churn_df['last_date_of_month_6'] - churn_df['date_of_last_rech_data_6']).dt.days
churn_df['num_of_days_last_rech_data_7'] = (churn_df['last_date_of_month_7'] - churn_df['date_of_last_rech_data_7']).dt.days
churn_df['num_of_days_last_rech_data_8'] = (churn_df['last_date_of_month_8'] - churn_df['date_of_last_rech_data_8']).dt.days

print(churn_df.shape)
churn_df.head()

(29953, 186)


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0      7000842753        109             0.0             0.0             0.0   
7      7000701601        109             0.0             0.0             0.0   
8      7001524846        109             0.0             0.0             0.0   
21     7002124215        109             0.0             0.0             0.0   
23     7000887461        109             0.0             0.0             0.0   

   last_date_of_month_6 last_date_of_month_7 last_date_of_month_8    arpu_6  \
0            2014-06-30           2014-07-31           2014-08-31   197.385   
7            2014-06-30           2014-07-31           2014-08-31  1069.180   
8            2014-06-30           2014-07-31           2014-08-31   378.721   
21           2014-06-30           2014-07-31           2014-08-31   514.453   
23           2014-06-30           2014-07-31           2014-08-31    74.350   

      arpu_7  ...  total_rech_call_plus_data_7  total_rech_call_plus_data_8  \
0    214.816  ...                        504.0                        504.0   
7   1349.850  ...                        790.0                       3638.0   
8    492.223  ...                        955.0                        327.0   
21   597.753  ...                        680.0                        718.0   
23   193.897  ...                       1166.0                        979.0   

    total_rech_call_plus_data_good_phase  Churn  num_of_days_last_rech_6  \
0                                  559.0      1                      9.0   
7                                 1185.0      1                      3.0   
8                                  696.0      0                      5.0   
21                                 640.0      0                      0.0   
23                                 583.0      0                     12.0   

    num_of_days_last_rech_7  num_of_days_last_rech_8  \
0                      15.0                     23.0   
7                       6.0                      5.0   
8                       0.0                      1.0   
21                      0.0                      0.0   
23                     24.0                      7.0   

    num_of_days_last_rech_data_6  num_of_days_last_rech_data_7  \
0                            9.0                          15.0   
7                            NaN                           NaN   
8                            NaN                           0.0   
21                           NaN                           NaN   
23                           NaN                          24.0   

    num_of_days_last_rech_data_8  
0                           23.0  
7                            NaN  
8                            8.0  
21                           NaN  
23                          25.0  

[5 rows x 186 columns]